mp2-v1/src/values_extraction/planner.rs

```diff
@@ -0,0 +1,430 @@
+
+        let update_tree = UpdateTree::<B256>::from_paths(key_paths, epoch as i64);
+
+        let mut update_plan = update_tree.clone().into_workplan();
+
+        while let Some(Next::Ready(work_plan_item)) = update_plan.next() {
+            let node_type = data_store
+                .get(work_plan_item.k())
+                .ok_or(anyhow::anyhow!(
+                    "No ProofData found for key: {:?}",
+                    work_plan_item.k()
+                ))?
+                .node_type;
+
+            let update_tree_node =
+                update_tree
+                    .get_node(work_plan_item.k())
+                    .ok_or(anyhow::anyhow!(
+                        "No UpdateTreeNode found for key: {:?}",
+                        work_plan_item.k()
+                    ))?;
+
+            match node_type {
+                NodeType::Leaf => {
+                    let proof_data =
+                        data_store
+                            .get_mut(work_plan_item.k())
+                            .ok_or(anyhow::anyhow!(
+                                "No ProofData found for key: {:?}",
+                                work_plan_item.k()
+                            ))?;
+                    let input = CircuitInput::new_receipt_leaf(
+                        &proof_data.node,
+                        proof_data.tx_index.unwrap(),
+                        self,
+                    );
+                    let proof = generate_proof(pp, input)?;
+                    proof_data.proof = Some(proof);
+                    update_plan.done(&work_plan_item)?;
+                }
+                NodeType::Extension => {
+                    let child_key = update_tree.get_child_keys(update_tree_node);
+                    if child_key.len() != 1 {
+                        return Err(anyhow::anyhow!("When proving extension node had {} many child keys when we should only have 1", child_key.len()));
+                    }
+                    let child_proof = data_store
+                        .get(&child_key[0])
+                        .ok_or(anyhow::anyhow!(
+                            "Extension node child had no proof data for key: {:?}",
+                            child_key[0]
+                        ))?
+                        .clone();
+                    let proof_data =
+                        data_store
+                            .get_mut(work_plan_item.k())
+                            .ok_or(anyhow::anyhow!(
+                                "No ProofData found for key: {:?}",
+                                work_plan_item.k()
+                            ))?;
+                    let input = CircuitInput::new_extension(
+                        proof_data.node.clone(),
+                        child_proof.proof.ok_or(anyhow::anyhow!(
+                            "Extension node child proof was a None value"
+                        ))?,
+                    );
```

Comment:
Maybe an update tree of `proof_data` would work and then DQ can insert the children proofs when it has them to build the `CircuitInput` that it would then pass down to a worker ?